arxiv-opensearch-cohere/README.md

```diff
@@ -0,0 +1,256 @@
+# Cohere / Opensearch implementation
+
+The following steps outline how a user would perform vector search with Cohere and Opensearch. Starting with spinning up a local OpenSearch instance and how a user would store Cohere embeddings in an OpenSearch index, and then explaining the methods for retrieval using these embeddings.
+Note, these instructions are based on OpenSearch v2.5.0.
+
+## Step 1: Spin up an instance of OpenSearch
+
+Run `docker-compose up` to spin up the Opensearch instance running at `http://localhost:9200`
+
+To make sure your server is running you can run `curl localhost:9200` and you should see a similar output:
+
+```
+{
+  "name" : "c77963d46f00",
+  "cluster_name" : "docker-cluster",
+  "cluster_uuid" : "Qn6HcFspROKuaMlmKbYoQQ",
+  "version" : {
+    "distribution" : "opensearch",
+    "number" : "2.5.0",
+    "build_type" : "tar",
+    "build_hash" : "b8a8b6c4d7fc7a7e32eb2cb68ecad8057a4636ad",
+    "build_date" : "2023-01-18T23:49:00.584806002Z",
+    "build_snapshot" : false,
+    "lucene_version" : "9.4.2",
+    "minimum_wire_compatibility_version" : "7.10.0",
+    "minimum_index_compatibility_version" : "7.0.0"
+  },
+  "tagline" : "The OpenSearch Project: https://opensearch.org/"
+}
+```
+
+
+## Step 2: Embed your documents 
+We will be utilizing Cohere's `embed` endpoint to embed documents into vectors. For demo purposes, we will be using the Cohere `small` model which will output vectors with 1024 floating point elements. 
+
+To follow along, download the [arxiv dataset](https://www.kaggle.com/datasets/Cornell-University/arxiv). We subset it to 5000 rows for example purposes and saved it to a local `/data` folder. 
+
+You'll need to create an account at [cohere.ai](https://dashboard.cohere.ai/) and grab your free trial `API_KEY`. 
+
+Use our Python package [cohere](https://pypi.org/project/cohere/) to create a client to interact with our embed endpoint. You can then instantiate a client like so:
+```python
+import cohere
+co = cohere.Client(<YOUR_API_KEY>)
+```
+
+Read in your local data file and create a list of texts to send to the embed endpoint. In this example, we will create embeddings using the `abstract` column.  
+```python
+import pandas as pd
+
+df = pd.read_csv(<PATH_TO_DATASET>).fillna("").reset_index(drop=True)
+
+texts = []
+for text in df["abstract"].values.tolist():
+    texts.append(text[0])
+```
+
+Once you have a list of texts to embed, hit the `embed` endpoint using the following helper function:
+```python
+import numpy as np 
+
+def get_cohere_embedding(text: Union[str, List[str]]) -> List[float]:
+    """
+    Embed a single text with cohere client and return list of float32
+    """
+    if type(text) == str:
+        embed = np.array(co.embed([text], model=COHERE_MODEL).embeddings)
```

Comment:
not clear where COHERE_MODEL comes from here.  the text above only mentions 'small' but now that it goes here?

Reply:
this is a snippet of the helper function to call; I was intending to give the user an idea of how things work and then the last line is the "full script can be run using `python cache_vectors.py`" so the configs are imported in that script from a global [config](https://github.com/cohere-ai/samples/blob/f2cbd376340237e654d7c217bab858f0e3bd76cb/arxiv-opensearch-cohere/config.py#L10) file. 
Do you think that's too confusing? I could make it a parameter to pass to the function instead

Reply:
actually, its probably not the best. I've changed the `get_cohere_embeddings` function to accept a `model_name`